stan/math/rev/mat/functor/kinsol_solve.hpp

```diff
@@ -0,0 +1,126 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_KINSOL_SOLVE_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_KINSOL_SOLVE_HPP
+
+#include <stan/math/prim/mat/fun/to_array_1d.hpp>
+#include <stan/math/prim/mat/fun/to_vector.hpp>
+#include <stan/math/rev/mat/functor/kinsol_data.hpp>
+#include <stan/math/rev/mat/functor/algebra_system.hpp>
+#include <stan/math/prim/mat/err/check_flag.hpp>
+
+#include <kinsol/kinsol.h>
+#include <sunmatrix/sunmatrix_dense.h>
+#include <sunlinsol/sunlinsol_dense.h>
+#include <nvector/nvector_serial.h>
+
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the solution to the specified algebraic system,
+ * given an initial guess. Invokes the Kinsol solver from Sundials.
+ *
+ * @tparam F type of equation system function.
+ * @tparam T type of initial guess vector.
+ * @param[in] f Functor that evaluated the system of equations.
+ * @param[in] x Vector of starting values.
+ * @param[in] y Parameter vector for the equation system. The function
+ *            is overloaded to treat y as a vector of doubles or of a
+ *            a template type T.
+ * @param[in] dat Continuous data vector for the equation system.
+ * @param[in] dat_int Integer data vector for the equation system.
+ * @param[in, out] msgs The print stream for warning messages.
+ * @param[in] scaling_step_tol Scaled-step stopping tolerance. If
+ *            a Newton step is smaller than the scaling step
+ *            tolerance, the code breaks, assuming the solver is no
+ *            longer making significant progress (i.e. is stuck)
+ * @param[in] function_tolerance determines whether roots are acceptable.
+ * @param[in] max_num_steps Maximum number of function evaluations.
+ * @param[in] custom_jacobian If 0, use Kinsol's default to compute the
+ *            jacobian for the Newton step, namely Quotient Difference
+ *            (finite difference). If 1, use reverse-mode AD, unless
+ *            the user specifies their own method.
+ * @param[in] J_f A functor that computes the Jacobian for the Newton step.
+ *            Defaults to reverse-mode AD.
+ * @param[in] steps_eval_jacobian Maximum number of steps before the
+ *            Jacobian gets recomputed. Note that Kinsol's default is 10.
+ *            If equal to 1, the algorithm computes exact Newton steps.
+ * @param[in] global_line_search does the solver use a global line search?
+ *            If equal to KIN_NONE, no, if KIN_LINESEARCH, yes.
+ * @return x_solution Vector of solutions to the system of equations.
+ * @throw <code>std::invalid_argument</code> if Kinsol returns a negative
+ *        flag when setting up the solver.
+ * @throw <code>boost::math::evaluation_error</code> if Kinsol returns a
+ *        negative flag after attempting to solve the equation.
+ */
+template <typename F1, typename F2 = kinsol_J_f>
+Eigen::VectorXd kinsol_solve(
```

Comment:
there are simply too many control parameters that may affect the solver behavior, especially consider immediately after this I'd like to add Picard solver in KINSOL(see earlier comment on solver name). So it's a good idea to add a `struct kinsol_control` dedicated for solver control & options. This way in the future the flow in `kinsol_solve` could remain the same even if we decide to add/change controls.

Reply:
Your idea is to pass an argument `kinsol_control` to the solver, instead of individual arguments for each tuning option, right? While this simplifies the signature of `kinsol_solve`, it burdens us with a new structure, that will need to be edited as we add new controls. I don't see the benefit of doing this; except perhaps that you can give the structure different constructors and thereby gain some flexibility on which arguments are optional.

Reply:
As we add Picard solver and possibly sparse solver, this control param thing will likely get out of control. What I'm proposing is to preserve `kinsol_solve` 's solution flow, as it maps to the `skeleton` in the kinsol user guide, and have a separate class to provide algorithm controls, in line with https://refactoring.guru/smells/long-parameter-list

Reply:
I'm still not quite convinced. It'd be helpful to see how we would take advantage of this when we add the Picard and sparse solver. For now, here's what I'm imagining:
```
str kinsol_control{
  double scaling_step;
  double function_tolerance;
  long int max_num_steps;
  . . .

  set_control(void* kinsol_memory) {
    check_flag_sundials(KINSetNumMaxIters(kinsol_memory, max_num_steps),
                        "KINSetNumMaxIters");
    . . .
}
```
Then we pass a `kinsol_control` object to `kinsol_solve()`.

Reply:
it's a blocker for this review, so feel free to move on.

Reply:
I wouldn't say it's a blocker. I do want to better understand what you have in mind (my presumption being that it's a good idea).

Reply:
Sorry, I meant to say it's not a show-stopper for this PR. So I'm fine we let it play out in the (near) future when I work on Picard solver.

Reply:
ok. I'm all for step-by-step PRs.